tensorflow_model_optimization/python/core/sparsity_tf2/riglpruner.py

```diff
@@ -0,0 +1,337 @@
+        in rank-2 weight tensors.
+      block_pooling_type: (optional) The function to use to pool weights in the
+        block. Must be 'AVG' or 'MAX'.
+      initializer: the initial sparsity distribution Callable for each layer of the network.
+      stateless: whether or not being run on TPU/multi-GPU workers.
+      seed: assigned by PruningConfig for multiworker consistency in random processes.
+      seed_offset: added to seed to run different experiments.
+      reinit: boolean for whether to reinitialize a connection that was drop and regorwn to 
+        its original value, or to set it to 0
+    """
+    super(RiGLPruner, self).__init__(
+        update_schedule, block_size, block_pooling_type)
+    self.initializer = initializer
+    self.target_sparsity = sparsity
+    self.update_schedule = update_schedule
+    self._block_size = block_size
+    self.block_pooling_type = block_pooling_type
+    self._stateless = stateless
+    self._seed = seed
+    self._seed_offset = seed_offset
+    self._noise_std = noise_std
+    self._reinit_when_same = reinit
+  
+  def create_slots(self, optimizer, var):
+    base_dtype = var.dtype
+    optimizer.add_slot(var, 'mask', initializer=self.initializer(self.target_sparsity))
+
+
+  def _validate_block(self, var):
+    if self._block_size != [1, 1]:
+      if var.get_shape().ndims != 2:
+        raise ValueError('Block Sparsity can only be used for layers which '
+                          'have 2-dimensional weights.')
+
+  def _random_uniform(self, step, *args, **kwargs):
+    """Uniform noise distribution"""
+    if self._stateless:
+      offset_seed = self._seed_offset + kwargs.get('seed', 0)
+      kwargs['seed'] = tf.cast(
+        tf.stack([offset_seed, step], tf.int32)
+      )
+      return tf.random.stateless_uniform(*args, **kwargs)
+    return tf.random.uniform(*args, **kwargs)
+
+  def _random_normal(self, step, *args, **kwargs):
+    """Gaussian noise distribution"""
+    if self._stateless:
+      offset_seed = self._seed_offset + kwargs.get('seed', 0)
+      kwargs['seed'] = tf.cast(
+        tf.stack([offset_seed, step], tf.int32)
+      )
+      return tf.random.stateless_normal(*args, **kwargs)
+    return tf.random.normal(*args, **kwargs)
+
+  def _get_grow_grads(self, optimizer, step, mask, grad):
+    """Grow connections baased on gradient information.
+    """
+    grad_scores = tf.math.abs(grad)
+    return grad_scores
+
+  def _get_drop_weights(self, optimizer, step, mask, weight, noise_std=0):
+    """
+    Drop connections based on weight magnitude.
+    """
+    masked_weight = mask * weight
+    weight_scores = tf.math.abs(masked_weight)
+    # add noise to break ties, although the possibility is very low
+    if noise_std != 0:
+      weight_scores.assign_add(self._random_normal(
+        weight_scores.shape, stddev=noise_std, dtype=weight_scores.dtype,
+        seed=(hash(weight.name + 'drop'))))
```

Comment:
It's probably better to just use `self._seed` as the seed, we'd like to avoid having weight names be load-bearing in terms of the random number generation.